backend/src/apiserver/resource/resource_manager.go

```diff
@@ -493,58 +493,105 @@ func (r *ResourceManager) DeleteJob(jobID string) error {
 func (r *ResourceManager) ReportWorkflowResource(workflow *util.Workflow) error {
 	if _, ok := workflow.ObjectMeta.Labels[util.LabelKeyWorkflowRunId]; !ok {
 		// Skip reporting if the workflow doesn't have the run id label
-		return nil
+		return util.NewInvalidInputError("Workflow missing the Run ID label")
 	}
 	runId := workflow.ObjectMeta.Labels[util.LabelKeyWorkflowRunId]
 	jobId := workflow.ScheduledWorkflowUUIDAsStringOrEmpty()
 
+	if workflow.PersistedFinalState() {
+		// If workflow's final state has being persisted, the workflow should be garbage collected.
+		err := r.workflowClient.Delete(workflow.Name, &v1.DeleteOptions{})
```

Comment:
It's confusing to me that the wf will be deleted once it's in final state. However, read it more, it appears PersistendFinalState is different from FinalState.

Reply:
yeah PersistendFinalState = FinalState + Workflow final state being persisted.